### PR TITLE
Add geometry shaders and refactor shader objects

### PR DESCRIPTION
**Pull request by djhoese**
Resolved merge of https://github.com/vispy/vispy/pull/1316 and current `master`. See that PR for more information.

### CODE REVIEW

**Comment by davidh-ssec** [comment]
> once the CIs are happy

Yeah that's the tough part. ;)

**Comment by davidh-ssec** [comment]
@campagnola If you have time could you try to pop in to the gitter chat every once and a while, at least until this PR gets merged.

**Comment by djhoese** [comment]
@campagnola @larsoner I mentioned this on gitter but gitter is down right now so I'm not sure if you've seen it. What do you guys get with this branch and running `examples/basics/gloo/geometry_shader.py`. On travis it fails because travis doesn't support OpenGL above 1.30. On my OSX machine I get a framebuffer error. I'm not sure if this is something wrong with the geometry shader or if this is related to the OSX framebuffer warnings I've seen before. With a lot of vispy examples on OSX High Sierra I get framebuffer errors for the first draw but then after that everything is fine. With regular gl (default GL backend) I get a warning (I think), with gl+ I get an error:

```
$ python examples/basics/gloo/geometry_shader.py 
WARNING: Traceback (most recent call last):
  File "examples/basics/gloo/geometry_shader.py", line 87, in <module>
    c = Canvas()
  File "examples/basics/gloo/geometry_shader.py", line 76, in __init__
    self.show()
  File "/Users/davidh/repos/git/vispy/vispy/app/canvas.py", line 429, in show
    self._backend._vispy_set_visible(visible)
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 352, in _vispy_set_visible
    self.showNormal()
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 448, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 720, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/latebind.py", line 41, in __call__
    return self._finalCall( *args, **named )
TypeError: 'NoneType' object is not callable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "examples/basics/gloo/geometry_shader.py", line 83, in on_draw
    self.program.draw('points')
  File "/Users/davidh/repos/git/vispy/vispy/gloo/program.py", line 529, in draw
    canvas.context.flush_commands()
  File "/Users/davidh/repos/git/vispy/vispy/gloo/context.py", line 171, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 224, in flush
    self._shared.flush(parser)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 146, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 468, in parse
    self._parse(command)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 436, in _parse
    ob.set_data(*args)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 567, in set_data
    errors = gl.glGetShaderInfoLog(self._handle)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/gl/_pyopengl2.py", line 131, in glGetShaderInfoLog
    res = GL.glGetShaderInfoLog(shader)
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/latebind.py", line 61, in __call__
    return self.wrapperFunction( self.baseFunction, *args, **named )
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/GL/VERSION/GL_2_0.py", line 320, in glGetShaderInfoLog
    length = int(glGetShaderiv(obj, GL_INFO_LOG_LENGTH))
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/latebind.py", line 61, in __call__
    return self.wrapperFunction( self.baseFunction, *args, **named )
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/GL/VERSION/GL_2_0.py", line 271, in glGetShaderiv
    shader, pname, status
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/latebind.py", line 45, in __call__
    return self._finalCall( *args, **named )
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/wrapper.py", line 664, in wrapperCall
    raise err
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/wrapper.py", line 657, in wrapperCall
    result = wrappedOperation( *cArguments )
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/platform/baseplatform.py", line 402, in __call__
    return self( *args, **named )
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/error.py", line 232, in glCheckError
    baseOperation = baseOperation,
OpenGL.error.GLError: GLError(
	err = 1286,
	description = b'invalid framebuffer operation',
	baseOperation = glGetShaderiv,
	pyArgs = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	),
	cArgs = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	),
	cArguments = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	)
)
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt4) at 0x102eb6668>> for DrawEvent
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt4) at 0x102eb6668>> repeat 2
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt4) at 0x102eb6668>> repeat 4
```

**Comment by campagnola** [comment]
The example runs on my system (ubuntu / nVidia).

**Comment by djhoese** [comment]
Very interesting @campagnola. I think it is all a bug in OSX drivers. Even when I run the `gpuimage.py` example I get:

```
$ python examples/basics/gloo/gpuimage.py --vispy-log=debug
vispy.app.application:Application._use(251): DEBUG: Selected backend _pyqt4
vispy.gloo.program:Program.draw(522): DEBUG: Program drawing 'triangle_strip' with (0, 4)
vispy.util.logs:log_exception(271): WARNING: Traceback (most recent call last):
  File "examples/basics/gloo/gpuimage.py", line 113, in <module>
    canvas = Canvas()
  File "examples/basics/gloo/gpuimage.py", line 99, in __init__
    self.show()
  File "/Users/davidh/repos/git/vispy/vispy/app/canvas.py", line 429, in show
    self._backend._vispy_set_visible(visible)
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 352, in _vispy_set_visible
    self.showNormal()
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 448, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 720, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "examples/basics/gloo/gpuimage.py", line 110, in on_draw
    self.program.draw('triangle_strip')
  File "/Users/davidh/repos/git/vispy/vispy/gloo/program.py", line 529, in draw
    canvas.context.flush_commands()
  File "/Users/davidh/repos/git/vispy/vispy/gloo/context.py", line 171, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 224, in flush
    self._shared.flush(parser)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 146, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 469, in parse
    self._parse(command)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 429, in _parse
    ob.draw(*args)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 979, in draw
    self._pre_draw()
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 930, in _pre_draw
    self._validate()
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 942, in _validate
    % gl.glGetProgramInfoLog(self._handle))
RuntimeError: Program validation error:
Validation Failed: Current draw framebuffer is invalid.
vispy.util.logs:_handle_exception(325): ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt4) at 0x1021ab550>> for DrawEvent
```

**Comment by djhoese** [comment]
But note that the gpuimage example keeps drawing due to the timer and all of the future draw operations succeed. I think it is some framebuffer initialization issue with OSX.

**Comment by djhoese** [comment]
I get this error specifically with the geometry shader example on OSX High Sierra and even resizing doesn't fix it:

```
$ python examples/basics/gloo/geometry_shader.py 
WARNING: Traceback (most recent call last):
  File "examples/basics/gloo/geometry_shader.py", line 87, in <module>
    c = Canvas()
  File "examples/basics/gloo/geometry_shader.py", line 76, in __init__
    self.show()
  File "/Users/davidh/repos/git/vispy/vispy/app/canvas.py", line 429, in show
    self._backend._vispy_set_visible(visible)
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 352, in _vispy_set_visible
    self.showNormal()
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 448, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 720, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "examples/basics/gloo/geometry_shader.py", line 83, in on_draw
    self.program.draw('points')
  File "/Users/davidh/repos/git/vispy/vispy/gloo/program.py", line 529, in draw
    canvas.context.flush_commands()
  File "/Users/davidh/repos/git/vispy/vispy/gloo/context.py", line 171, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 224, in flush
    self._shared.flush(parser)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 146, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 469, in parse
    self._parse(command)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 437, in _parse
    ob.set_data(*args)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 568, in set_data
    errors = gl.glGetShaderInfoLog(self._handle)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/gl/_pyopengl2.py", line 131, in glGetShaderInfoLog
    res = GL.glGetShaderInfoLog(shader)
  File "latebind.pyx", line 44, in OpenGL_accelerate.latebind.Curry.__call__ (src/latebind.c:1201)
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/GL/VERSION/GL_2_0.py", line 320, in glGetShaderInfoLog
    length = int(glGetShaderiv(obj, GL_INFO_LOG_LENGTH))
  File "latebind.pyx", line 44, in OpenGL_accelerate.latebind.Curry.__call__ (src/latebind.c:1201)
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/GL/VERSION/GL_2_0.py", line 271, in glGetShaderiv
    shader, pname, status
  File "latebind.pyx", line 32, in OpenGL_accelerate.latebind.LateBind.__call__ (src/latebind.c:989)
  File "wrapper.pyx", line 318, in OpenGL_accelerate.wrapper.Wrapper.__call__ (src/wrapper.c:6561)
  File "wrapper.pyx", line 311, in OpenGL_accelerate.wrapper.Wrapper.__call__ (src/wrapper.c:6439)
  File "/Users/davidh/anaconda/envs/vispy_test_36/lib/python3.6/site-packages/OpenGL/platform/baseplatform.py", line 402, in __call__
    return self( *args, **named )
  File "errorchecker.pyx", line 53, in OpenGL_accelerate.errorchecker._ErrorChecker.glCheckError (src/errorchecker.c:1218)
OpenGL.error.GLError: GLError(
	err = 1286,
	description = b'invalid framebuffer operation',
	baseOperation = glGetShaderiv,
	pyArgs = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	),
	cArgs = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	),
	cArguments = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	)
)
```

**Comment by djhoese** [comment]
Progress report: I hate OSX (warning: this is going to be a long comment). I'm going to put every error message I have stumbled upon so that someone might find this PR when googling.

**TL;DR: OpenGL >2.1 is not fully supported by VisPy on OSX at this time. Supporting it would require modifying VisPy to support the requires of GL Core profiles (VAOs, etc).**

The last step to this PR as far as desktop compatibility goes is making sure it works for me locally by using the geometry_shader.py example. When I run this example with everything at its defaults and using the PyQt4 backend and using the examples forced GL backend of `gl+`, I get:

<details>
<summary>PyQt4 geometry_shader.py Error</summary>

```
OpenGL.error.GLError: GLError(
	err = 1286,
	description = b'invalid framebuffer operation',
	baseOperation = glGetShaderiv,
	pyArgs = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	),
	cArgs = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	),
	cArguments = (
		2,
		GL_INFO_LOG_LENGTH,
		array([50], dtype=int32),
	)
)
```

</details>
<br/>

Similarly when running the hello_fbo.py example I get this on OSX PyQt4:

<details>
<summary>PyQt4 hello_fbo.py Error</summary>

```
WARNING: Traceback (most recent call last):
  File "examples/basics/gloo/hello_fbo.py", line 120, in <module>
    c = Canvas()
  File "examples/basics/gloo/hello_fbo.py", line 98, in __init__
    self.show()
  File "/Users/davidh/repos/git/vispy/vispy/app/canvas.py", line 429, in show
    self._backend._vispy_set_visible(visible)
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 354, in _vispy_set_visible
    self.showNormal()
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 450, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/davidh/repos/git/vispy/vispy/app/backends/_qt.py", line 722, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 455, in __call__
    self._invoke_callback(cb, event)
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 475, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/Users/davidh/repos/git/vispy/vispy/util/event.py", line 471, in _invoke_callback
    cb(event)
  File "examples/basics/gloo/hello_fbo.py", line 116, in on_draw
    self._program2.draw('triangle_strip')
  File "/Users/davidh/repos/git/vispy/vispy/gloo/program.py", line 529, in draw
    canvas.context.flush_commands()
  File "/Users/davidh/repos/git/vispy/vispy/gloo/context.py", line 171, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 224, in flush
    self._shared.flush(parser)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 146, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 469, in parse
    self._parse(command)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 429, in _parse
    ob.draw(*args)
  File "/Users/davidh/repos/git/vispy/vispy/gloo/glir.py", line 963, in draw
    gl.check_error('Check before draw')
  File "/Users/davidh/repos/git/vispy/vispy/gloo/gl/__init__.py", line 209, in check_error
    raise err
RuntimeError: OpenGL got errors (Check before draw): GL_INVALID_FRAMEBUFFER_OPERATION
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt4) at 0x115d70278>> for DrawEvent
```

</details>
<br/>

I get this warning for just about every vispy example (except for only the very basic examples like creating a context). So I figured this was an issue with PyQt4 or OSX. So I tried a PyQt5 backend and `hello_fbo.py` gives no warnings. I then tried the `geometry_shader.py` example with PyQt5:

<details>
<summary>PyQt5 geometry_shader.py Error</summary>

```
RuntimeError: Shader compilation error in 35633:
    on line 1: '' :  version '330' is not supported
      #version 330
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt5) at 0x106d31c88>> for DrawEvent
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt5) at 0x106d31c88>> repeat 2
```

</details>
<br/>

So this is when I discovered that Qt and OSX default to using OpenGL 2.1:

https://github.com/vispy/vispy/issues/952
https://github.com/vispy/vispy/issues/1157
https://github.com/vispy/vispy/issues/1138

The first issue above mentions a quick hack to try to force Qt to use 4.1 (or other), but OSX requires that you use the Core profile.

```python
# vispy/app/backends/_qt.py

def _set_config(c):
    """Set the OpenGL configuration"""
    glformat = QGLFormat()
    glformat.setVersion(4, 1)
    glformat.setProfile(QGLFormat.CoreProfile)
    # and so on...
```

Using the Core profile causes VisPy to fail due to some GL 2 stuff that it's doing. Mainly it has no concept of [VertexArrayObjects](https://www.khronos.org/opengl/wiki/Tutorial2:_VAOs,_VBOs,_Vertex_and_Fragment_Shaders_(C_/_SDL)).

I tried a few things to get it to work, but without fully implementing VAOs I'm not sure there is much I can do on OSX. I'm also not sure if using VAOs is GL 2.0 or GL ES 2.0 compatible or if it can be completely handled by the parsing side of GLIR instead of on the visual/gloo side of things. I was hoping to add Tessellation shaders after getting this PR to work, but that might have to wait until I can OpenGL >2.0 to work on OSX.

**Comment by djhoese** [comment]
@campagnola I think I'm going to make the `DATA` command for the shaders match the DATA command for everything else. That means it will have an unused offset of 0. This way I don't have to do any special handling of the DATA commands in the notebook backend. It also means that the strings will be sent to the notebook as binary buffers which doesn't do anything for us, but I figure it'll help us in the future if we ever need to compress buffer data or shaders for some reason.

**Comment by djhoese** [comment]
I fixed and changed a lot of stuff in vispy.js to make this PR work last time I worked on it. To reduce wasted effort I'm combining #1601 in to this PR. @larsoner you approved this PR in the past. Assuming I can get tests to pass I think I'd like to merge this before 0.6. I'm sure @campagnola would be happy about that too.

This PR does make some major low-level changes though which I can see people not being happy about if they got used to "hacking" things in vispy.

**Comment by djhoese** [comment]
I'm going to merge this once tests pass. It has gotten way to large but each commit is pretty important. I forgot to skip the size checks in the last commits so the tests may show failing but I will only merge if the actual tests pass.

**Comment by djhoese** [comment]
Failures are only size of the commit issues. Merging...
